### ISSUE TITLE
az login on headless vm opens w3m due to a regression.

### ISSUE THREAD

**Issue by temporaer** [Azure/azure-cli#21112]
**Describe the bug**
In a headless linux VM, doing `az login` opens www-browser and that in turn opens w3m, a text-only browser that does not support the device login flow.

**To Reproduce**
ssh into headless VM without X forwarding. run `az login`.

**Expected behavior**
Device code flow should get initiated, ie, i should see a URL and a code to enter there after opening it in a javascript-capable browser.

**Environment summary**
I used the script install on linux to get the newest version.

**Additional context**
This is a regression, which was introduced by [this commit](https://github.com/Azure/azure-cli/commit/1b4c85187cf0a74e4192e25f514f3c27edeef967). Ie, it doesn't check DISPLAY envvar anymore.

**Comment by temporaer** [comment]
Note: since az cli code gets vendored by azureml, the same issue occurs in the azureml sdk's `InteractiveLoginAuthentication`, but can't be worked around using the `--use-device-code` flag.

**Comment by yonzhan** [comment]
@jiasli for awareness

**Comment by jiasli** [comment]
This is indeed a tricky topic. Previously in ADAL-based Azure CLI, we did prevent launching a `www-browser` browser:

https://github.com/Azure/azure-cli/blob/0a1edd5b0960f02162c89abaa97d9ab063d202ba/src/azure-cli-core/azure/cli/core/util.py#L710

And we use `powershell.exe` to launch the Windows host browser on WSL:

https://github.com/Azure/azure-cli/blob/0a1edd5b0960f02162c89abaa97d9ab063d202ba/src/azure-cli-core/azure/cli/core/util.py#L654-L661

But later on, WSL supports `www-browser` which launches the Windows host browser so MSAL uses `www-browser` to do auth code flow. The tricky part is that we can't know whether `www-browser` is a text browser or GUI browser:

- `www-browser` is a text browser on some Linux distributions
- `www-browser` is Windows host browser on WSL

@temporaer, is it possible that on some Linux distribution besides WSL `www-browser` is also a GUI browser?

**Comment by rayluo** [comment]
> But later on, WSL supports `www-browser` which launches the Windows host browser so MSAL uses `www-browser` to do auth code flow. The tricky part is that we can't know whether `www-browser` is a text browser or GUI browser:
> 
> * `www-browser` is a text browser on some Linux distributions
> * `www-browser` is Windows host browser on WSL

I concur with @jiasli's triage. In fact, there is another situation (other than WSL) that the `www-browser` would still work. That is the Azure Cloud Shell (although that one would require more ingredient - it would require Azure CLI to adjust its app registration a little bit - @jiasli & @yonzhan, we tested that proof-of-concept before.) Anyway, my point is "we can't know whether `www-browser` would eventually launch a real browser". So, the current approach is to still attempt the `www-browser`, but provide a hint for the user to manually switch to `az login --use-device-code` as plan B.

**Comment by temporaer** [comment]
I tend to disagree with the assessment from an outside point of view. It would be better to explicitly detect wsl or cloud shell rather than providing the confusing experience with w3m. Also in the vendored code in azureml, there's no way to use the flag you suggest since users may not even have az installed. The only way to fix that for me was to rename system executables.

I would rather have slightly more complex code than a bad user experience.

**Comment by temporaer** [comment]
One could also argue that this is a bug in wsl since it should provide x-www-browser, not www-Browser.

**Comment by jiasli** [comment]
> One could also argue that this is a bug in wsl since it should provide x-www-browser, not www-Browser

For this, do you mind opening an issue at https://github.com/microsoft/WSL/issues ?

**Comment by temporaer** [comment]
@jiasli i can but it won't solve the problem here.

**Comment by rayluo** [comment]
> **Describe the bug** In a headless linux VM, doing `az login` opens www-browser and that in turn opens w3m, a text-only browser that does not support the device login flow.
> 
> **To Reproduce** ssh into headless VM without X forwarding. run `az login`.
> 
> **Expected behavior** Device code flow should get initiated, ie, i should see a URL and a code to enter there after opening it in a javascript-capable browser.
> 
> ...
> 
> **Additional context** This is a regression, which was introduced by [this commit](https://github.com/Azure/azure-cli/commit/1b4c85187cf0a74e4192e25f514f3c27edeef967). Ie, it doesn't check DISPLAY envvar anymore.

Please allow me to take one step back to re-examine the situation here.

In this context, there are 2 methods for a user to sign-in:

1. Using a browser on the machine running your app (i.e. the `az`). This is the recommended approach, mainly because it is more user friendly, and more feature rich.
2. Usually, when/if there is no (capable?) browser on that machine, you can use a "device code flow" to complete the sign-in from a different machine.

The `az login --use-device-code` unambiguously uses method 2. The `az login` would generally attempt method 1, but historically it would automatically switch to method 2, when it believes method 1 would fail.

However, some traditionally "device-code only" scenarios are actually possible to still use method 1. For example, today the MSAL Python (which is the authentication library used by az 2.30+) is capable to facilitate a browser flow inside a local "text mode" docker container. Similarly, the scenario described by @temporaer , "ssh into headless VM without X forwarding", could also perform a browser-based login, when/if the ssh session also opens a tunnel beforehand. (I believe `az` does not currently expose these features; nonetheless it can be an intriguing feature request.)

In summary, `az login` in v2.30+ does not aggressively fallback to device code flow. Instead, we hope the users would build a new habit of "**if I want to use device code flow, I do `az login --use-device-code`. But if I do `az login` instead, I understand `az` will wait for either a browser flow's potential success, or wait for me to manually cancel and switch to device code flow**". That being said, we understand that the user experience is not ideal, especially during this transitional period.

**Comment by temporaer** [comment]
@rayluo thank you for the summary, I think the most critical bit is missing though: az login on the headless VM (or docker container!) will open a text mode browser that completely hides the hint to use the --use-device-code flag (it's running "full screen" in the terminal if you will). You have to first figure out how to exit that browser to see the hint. 

I'm not sure what this transitional period is you mention, can you explain? To me, it seems the user experience will stay bad unless the can_launch_browser method is fixed. 

Also note that contrary to your summary, the previous implementation of that function worked perfectly fine on WSL and opened a windows browser for login, *not* the device code flow.

**Comment by rayluo** [comment]
> I'm not sure what this transitional period is you mention, can you explain? To me, it seems the user experience will stay bad unless the can_launch_browser method is fixed.

By "transitional period", I was wishfully hoping users (especially new users) would eventually form an understanding of "`az login` and `az login --use-device-code` are 2 totally different login methods; it is my choice/duty to choose the right one to use". But right now, existing users (like you) already formed the habit of - and continue to expect - that running `az login` on VMs should automatically fall back to device code flow.

> the most critical bit is missing though: az login on the headless VM (or docker container!) will open a text mode browser that completely hides the hint to use the --use-device-code flag (it's running "full screen" in the terminal if you will). You have to first figure out how to exit that browser to see the hint.

Good point. Now I understand what you mean.

For what it's worth, previously when I tested the browser login in my docker container, my image has no browser available, not even a text browser. So, my proof-of-concept app was able to display a full URL which is usually clickable. The overall experience is a slick "one (extra) click to sign-in".

But if your environment defines a browser (or $BROWSER). That was supposed to be a signal to claim, "hey this system has a browser". Then there will be no easy way for any app to know how capable that "browser" is.

All in all, this is now a user experience topic. We can brainstorm. Azure CLI may choose to consider some environment's browser is questionable, and then decide to proactively display the `--use-device-code` hint BEFORE launching the browser; or Azure CLI may choose to fall back to the old way i.e. start the device code flow instead.

**Comment by temporaer** [comment]
This makes sense. If i may suggest a rule, www-browser should always be assumed to be a text browser, x-www-browser a fully capable browser. If cloud shell and wsl break this assumption, we can report this upstream and work around that broken behavior for existing systems by creating specific exceptions instead of making assuming all systems are broken in this way.

**Comment by temporaer** [comment]
@rayluo @jiasli we have more and more people getting hit by this issue. Can you reconsider providing a solution please?
